src/Cryptol/Parser/ExpandPropGuards.hs

```diff
@@ -0,0 +1,127 @@
+-- Stability   :  provisional
+-- Portability :  portable
+--
+-- Expands PropGuards into a top-level definition for each case, and rewrites
+-- the body of each case to be an appropriate call to the respectively generated
+-- function.
+--
+
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DeriveAnyClass #-}
+
+module Cryptol.Parser.ExpandPropGuards where
+
+import Cryptol.Parser.AST
+-- import Cryptol.Parser.Position (Range(..), emptyRange, start, at)
+-- import Cryptol.Parser.Names (namesP)
+import Cryptol.Utils.PP
+-- import Cryptol.Utils.Ident (mkIdent)
+import Cryptol.Utils.Panic (panic)
+-- import Cryptol.Utils.RecordMap
+
+import           Data.Text (pack)
+
+import GHC.Generics (Generic)
+import Control.DeepSeq
+
+-- | Monad
+
+type ExpandPropGuardsM a = Either Error a
+
+runExpandPropGuardsM :: ExpandPropGuardsM a -> Either Error a
+runExpandPropGuardsM m = m
+
+-- | Class
+class ExpandPropGuards a where
+  expandPropGuards :: a -> ExpandPropGuardsM a
+
+-- | Error
+data Error = NoSignature (Located PName)
+  deriving (Show,Generic, NFData)
+
+instance PP Error where
+  ppPrec _ err = case err of
+    NoSignature x -> 
+      text "At" <+> pp (srcRange x) <.> colon <+>
+      text "No signature provided for declaration that uses PROPGUARDS."
+
+-- | Instances
+
+instance ExpandPropGuards (Program PName) where
+  expandPropGuards (Program decls) = Program <$> expandPropGuards decls
+
+instance ExpandPropGuards (Module PName) where
+  expandPropGuards m = do
+    mDecls' <- expandPropGuards (mDecls m)
+    pure m { mDecls = mDecls' }
+
+instance ExpandPropGuards [TopDecl PName] where
+  expandPropGuards topDecls = concat <$> traverse f topDecls 
+    where 
+      f :: TopDecl PName -> ExpandPropGuardsM [TopDecl PName]
+      f (Decl topLevelDecl) = fmap mu <$> expandPropGuards [tlValue topLevelDecl]
+        where mu decl = Decl $ topLevelDecl { tlValue = decl }
+      f topDecl = pure [topDecl]
+
+instance ExpandPropGuards [Decl PName] where 
+  expandPropGuards decls = concat <$> traverse f decls 
+    where 
+      f (DBind bind) = fmap DBind <$> expandPropGuards [bind]
+      f decl = pure [decl]
+
+instance ExpandPropGuards [Bind PName] where
+  expandPropGuards binds = concat <$> traverse f binds
+    where 
+      f :: Bind PName -> Either Error [Bind PName]
+      f bind = case thing $ bDef bind of 
+        DPropGuards guards -> do
+          Forall params props t rng <-
+            case bSignature bind of
+              Just schema -> pure schema 
+              Nothing -> Left . NoSignature $ bName bind
+          let
+            g :: ([Prop PName], Expr PName) -> ExpandPropGuardsM (([Prop PName], Expr PName), Bind PName)
+            g (props', e) = do
+              bName' <- newName (bName bind) props'
+              -- call to generated function
+              let e' = foldr EApp (EVar $ thing bName') (patternToExpr <$> bParams bind)
```

Comment:
I think here you need to also apply the type arguments.  To see why, consider the following example:
```
isZero : {n} () -> Bit
isZero x
  | n == 0 => True
  | ()     => False
```
If you get a warning about "unused `n`" which is is not right.  Also, if you load with `:set debug=on` (it shows you the IR), you get the following:
```Main::isZero : {n} () -> Bit
Main::isZero =
  \{n} ->
    (propguard | n`903 ==
    0 => Main::isZero(n == 0) 0 <> |  => Main::isZero() 0)
```
so there seems to be some confusion with the quantifiers

Reply:
Ah yes, I see the problem. Usually it works when it can infer the right types from the arguments passed, but in this case it can't